cmd/mixtool/server.go

```diff
@@ -99,58 +98,47 @@ type ruleProvisioner struct {
 // to existing, does not provision them. It returns whether Prometheus should
 // be reloaded and if an error has occurred.
 func (p *ruleProvisioner) provision(r io.Reader) (bool, error) {
-	b := bytes.NewBuffer(nil)
-	tr := io.TeeReader(r, b)
-
-	f, err := os.Open(p.ruleFile)
-	if err != nil && !os.IsNotExist(err) {
-		return false, fmt.Errorf("open rule file: %w", err)
-	}
-	if os.IsNotExist(err) {
-		f, err = os.Create(p.ruleFile)
-		if err != nil {
-			return false, fmt.Errorf("create rule file: %w", err)
-		}
+	fileData, err := ioutil.ReadFile(p.ruleFile)
```

Comment:
Rather than reading both rule files completely into memory, i think it could be better to:

1. Write the new rules to a temp file
2. Get a file pointer to the old file
3. Compare the readers for both files using the existing helper fun
4. If equal, return early
5. If not equal, rename the temp file